src/interval.jl

```diff
@@ -338,7 +359,14 @@ function Base.merge(a::AbstractInterval, b::AbstractInterval)
 
     left = min(LeftEndpoint(a), LeftEndpoint(b))
     right = max(RightEndpoint(a), RightEndpoint(b))
-    return Interval(left, right)
+
+    # This promotion fixes the situation where one endpoint has a type of
+    # `InfExtended{T}` yet the ∞ value is of type `Infinite`.
+    # This will cause an error when trying to `promote(left, right)`
+    return Interval(
+        promote(left.endpoint, right.endpoint)...,
+        left.included, right.included
+    )
```

Comment:
Seems preferable to change the endpoint constructor (`Interval{T}(left::LeftEndpoint{T}, right::RightEndpoint{T})`)

Reply:
The problem here is that you'll end up with situations such that 
```
typeof(left) = Intervals.Endpoint{InfExtended{Int64},Intervals.Direction{:Left}()}
typeof(right) = Intervals.Endpoint{Int64,Intervals.Direction{:Right}()}
```
Thus you don't actually make it to that constructor you mention since the types are different. It ends up trying to call `Interval(f, l, inc...) = Interval(promote(f, l)..., inc...)` which fails because there are no rules on how to promote two endpoints.

Reply:
Would it make sense to also define a constructor for two endpoints where the types are different? I seems like we typically don't want to change what the Interval Type is, but with Infinity it seems to be required in certain situations

```
function Interval(left::LeftEndpoint{S}, right::RightEndpoint{D}) where {S, D}
    T = promote_type(S, D)
    return Interval{T}(left.endpoint, right.endpoint, left.included, right.included)
end
```

Reply:
That is one option. We may alternatively want to define our own `promote_rule`